File: smart-poultry-dashboard/src/api/pens.ts
import { api } from "./client";
import { Pen, PenCreate, PenUpdate } from "../types/pen";

export const listPens = () => api<Pen[]>("/pens");

export const getPen = (id: string) =>
  api<Pen>(`/pens/${id}`);

export const createPen = (data: PenCreate) =>
  api<Pen>("/pens", {
    method: "POST",
    body: JSON.stringify(data),
  });

export const updatePen = (id: string, data: PenUpdate) =>
  api<Pen>(`/pens/${id}`, {
    method: "PUT",
    body: JSON.stringify(data),
  });

export const deletePen = (id: string) =>
  api<Pen>(`/pens/${id}`, { method: "DELETE" });

File: app/api/pens.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.schemas.pen import PenCreate, PenUpdate, PenOut
from app.services.pen_service import create_pen, get_pen, get_pens, update_pen, delete_pen
from app.utils.db import get_db

router = APIRouter(tags=["Pens"], prefix="/api/pens")

@router.post("", response_model=PenOut)
def api_create_pen(payload: PenCreate, db: Session = Depends(get_db)):
    return create_pen(db, payload)

@router.get("", response_model=list[PenOut])
def api_list_pens(skip: int = 0, limit: int = 100, db: Session = Depends(get_db)):
    return get_pens(db, skip, limit)

@router.get("/{pen_id}", response_model=PenOut)
def api_get_pen(pen_id: str, db: Session = Depends(get_db)):
    return get_pen(db, pen_id)

@router.put("/{pen_id}", response_model=PenOut)
def api_update_pen(pen_id: str, payload: PenUpdate, db: Session = Depends(get_db)):
    return update_pen(db, pen_id, payload)

@router.delete("/{pen_id}", response_model=PenOut)
def api_delete_pen(pen_id: str, db: Session = Depends(get_db)):
    return delete_pen(db, pen_id)
